### ISSUE TITLE
Question: the path of SVFG for branches

### ISSUE THREAD

**Issue by Zhong0Lei** [SVF-tools/SVF#1306]
When using Saber to test this test code, the expected results were not obtained
my codes：
```
#include<stdlib.h>
typedef struct Node {
    int data;           
    struct Node *next;  
} Node;
int main() {
    Node* head1 = (Node*)malloc(sizeof(Node)); 
    Node* head2 = (Node*)malloc(sizeof(Node)); 
    int a = 1;
    int b = 1;
    Node *p = head1;
    if(a == b){
        p = head2;
    }
    free(p);
    return 0;
}
```
The correct result should be that both head1 and head2 report `PartialLeaf`

But the test results show that head1 has no memory leak, while head2 reports PartialLeak

![image](https://github.com/SVF-tools/SVF/assets/47477662/306441d1-d460-4625-8eca-7a9997c12df4)

I checked the slice diagram and found that the path from the malloc node of head1 to the free node did not pass through `if...then`  block, may I ask why this is?

**Comment by yuleisui** [comment]
@jumormt could you have a look?

**Comment by jumormt** [comment]
This case is tricky. Different from head2, the conditional free path for head1 should be the if-else branch. It seems that the value-flow guard between p@Line 11 and p@Line 15 should be `a != b` instead of a true value.

Regarding why head1 does not pass through the if-then block, the points-to set of head1@Line 11 does not pass to p@Line 13 (i.e., no indirect value-flow between p@Line 11 and p@Line 13) because it is a strong update at Line 13. Therefore, saber removes that redundant indirect value-flow.

**Comment by yuleisui** [comment]
@ZcoderL could you upload the c file and its svfg? @jumormt the strong update should remove 11->13 but should keep 11->15 and the value-flow guard should have a condition a!=b?

**Comment by jumormt** [comment]
Yes, 11->13 is removed due to a strong update. 11->15 is a normal direct def-use and is kept in SVFG. The source code with line number:

![image](https://github.com/SVF-tools/SVF/assets/44386076/c8d7e923-5253-48fe-a280-38b47b98a27c)

And the SVFG (relevant part):

![image](https://github.com/SVF-tools/SVF/assets/44386076/67340c0c-a67a-4bca-97f9-e0373c4e0c90)

The entire SVFG:

![svfg](https://github.com/SVF-tools/SVF/assets/44386076/62184723-ae50-4b1d-bc51-8e591ac24ae0)

**Comment by yuleisui** [comment]
The value-flow guard for 11-15 is missing? If not missing, there is supposed to be a conditional free?

**Comment by jumormt** [comment]
#1308 I have created a pull request. We need a heuristic to solve this case.

**Comment by yuleisui** [comment]
@ZcoderL could you try the recent merge and see whether it fixes your issue?
https://github.com/SVF-tools/SVF/pull/1309
